Pytest benchmark override with cpu and gpu times

MrRobot2211 opened the pull request:
The idea here is to override pytest_benchmark fixture calls to retain most of the funtionality but use `cupy.benchmark.repeat` for timings instead of `pytest_benchmark`'s internal `runner` object. This allows us to benchmark total, cpu and gpu times. I am striving for a fast solution, the sooner we can start benchmarking the better.

Maybe, this could done more elegantly by using a cutom `timer ` probably 
, but then one would run into trouble elsewhere in the code, and may de forced to create a wrapper calls for the oputput of the timer with an specialized `__add__` method and to change the library used for statistics. At that point registerin our own `pytest` plugin that links to `pytest_benchmark` maybe more suitable. That looks more like a long-term implementation for later on.


To quick test that the functionality works you should install pytest_benchmark  and do:

` python src/benchmark_tools/benchmark.py`

and you will be shown three different statistics: full time, cpu time and gpu time.

you can also build a json and see the separate statistics, this will allow it to play well  with Asier's benchmarking code.

Why it works
--------------
In the `Fixture` class the method `_add_stats`  appends a new benchmark (`bench_stats` of class `Metadata`) into a `BenchmarkSession` list.
 Each element of this list will then be taken one by one the `as_dict` method and dumped into a json file.
No runtime checks are performed in pytest_benchmark on the size of this list.
 Note that each bench_stats keepss the group and extra information from the `Fixture` wherein it was created.
 
By copying the structur in ` __call__`/ `pedantic` for our new funtion, iinstead of acreating a `bench_stats` with each call we will be creating three of them with different group attributes. Whose statistics will be calculated separately.

Pending tasks
----------------
- [x] Properly document the relevant behaviour changes from the original `pytest_benchmark` functionality.
- [x] Cleaning up and formatting the code 
- [x]  Add functionalitry to register the GPU name and features being used for the test. 


I am open to add more things to the list as the review progresses

Other similar packages
-----------------------------
- [rapids-pytest-benchmark](https://github.com/rapidsai/benchmark/tree/main/rapids_pytest_benchmark) implements a similar solution to ours for wrapping the `benchmark` module in `pytest-benchmark` it adds functionality for benchmarking memory use but it lacks `host` and `device` separate stats.
- [cupy-performance](https://github.com/cupy/cupy-performance) does something very similar to what we want but in a `unittest` style.

MrRobot2211 committed:
basic directory structure

MrRobot2211 committed:
adds basic implementation for wrapper

MrRobot2211 committed:
adds basic test functionality

MrRobot2211 committed:
corrects pedantioc cupy arguments

MrRobot2211 committed:
working first example

MrRobot2211 committed:
Merge branch 'main' of https://github.com/qutip/qutip-cupy into pytest_benchmark_override_with_cpu_and_gpu_times

hodgestar commented on benchmarks/cpu_gpu_times_wrapper.py:
```diff
@@ -0,0 +1,93 @@
+        else:
+            return orig_attr
+
+    def __setattr__(self, attr, value):
+        setattr(self.wrapped_class, attr, value)
```
I think we can safely remove `__setattr__`.

Setting `__setattr__` causes `self.x = y` to always call `__setattr__` and setting an attribute on self has to be done via `self.__dict__["x"] = y` (which is painful and easy to forget to do). On top of this, its hard to imagine why someone doing `self.x = y` on the wrapper object would want the attribute to be set on the wrapped *class*.

MrRobot2211 replied:
I need things to be set on the wrapped class because it is the one that does all the communications with the pytest benchmark modules. The wrapper class should not have any attributes besides the class being wrapped. Maybe there is an easier way to dinamically add this functionality on the classes ?

I think that the rapids version, which does not have the `__setattr__` may have issues when manually setting attributes , like in [Asier's code](https://github.com/qutip/qutip-tensorflow/blob/f2eba14d9010545deab1d300001375e72b15c456/tests/test_linear_algebra.py#L97), which I would like to reuse in part.

Notice that if we were to register our wrapper to `pytest` , when doing `benchmark.groups = "x" ` we wouldn't be setting the wrapped `groups`  attribute, which then gets pushed to the `BenchmarkSession`.

hodgestar commented:
Just linking to the source code of the pytest-benchmark class we're wrapping for those who are reading along and reviewing: https://github.com/ionelmc/pytest-benchmark/blob/master/src/pytest_benchmark/fixture.py

MrRobot2211 committed:
make pedanticupy arguments
compatible with pedantic

MrRobot2211 committed:
add benchmark  plot generation

MrRobot2211 committed:
add groups dynamically in benchmark tests

hodgestar commented:
@MrRobot2211 What is the status of this PR? Should I review again? Or would you like to make more changes first?

MrRobot2211 committed:
working with one benchmark

MrRobot2211 committed:
add device_id settings as settable
global variables. The structure had
to be changed for this to work
alternatively the structure could be
 kept if setup.cfg is changed to recognize the extra-module

MrRobot2211 committed:
fix benchmark device setup

MrRobot2211 committed:
add array creation by dtype

MrRobot2211 committed:
Merge branch 'main' of https://github.com/qutip/qutip-cupy into pytest_benchmark_override_with_cpu_and_gpu_times

MrRobot2211 committed:
aadd device half precision option

MrRobot2211 committed:
format

MrRobot2211 committed:
format

MrRobot2211 committed:
add benchmarks to setup.cfg

MrRobot2211 committed:
correct plot labeling

MrRobot2211 committed:
add extra-dependencies for benchmark

MrRobot2211 commented:
@hodgestar  I am missing here a tiny little code to generalize benchmarks to all operations, but I think it is a good idea that you take a look. I am not very comfortable about having promoted benchmark-tools to a module but did not see any better option.

hodgestar reviewed (commented):
I left a couple of comments. Do you have example output from the benchmark runs? Perhaps we can add that as part of this PR somehow.
